astropy/coordinates/_optimizations.pyx

```diff
@@ -0,0 +1,83 @@
+# cython: language_level = 3
+import numpy as np
+cimport numpy as np
+cimport cython
+from libc.math cimport isfinite as _isfinite, M_PI
+from libc.stdint cimport int32_t, int64_t
+
+
+ctypedef fused dtype_t:
+    int32_t
+    int64_t
+    float
+    double
+
+
+cdef bint isfinite(dtype_t val):
+    if dtype_t is int32_t or dtype_t is int64_t:
+        return True
+    else:
+        return _isfinite(val)
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.nonecheck(False)
+@cython.cdivision(True)
```

Comment:
To avoid repetition for all functions you could put those in a directive at the top of the file, e.g.:
```
# cython: boundscheck=False, nonecheck=False, wraparound=False, language_level=3, cdivision=True
```